signature/src/verifier.rs

```diff
@@ -0,0 +1,105 @@
+use crate::error::{ErrorKind, SignatureError};
+use crate::suite::{Ed25519Signature, Signature, VerificationRelation};
+
+pub trait DIDVerifier<S>
+where
+    S: Signature,
+{
+    fn verify(&self, msg: &[u8], signature: &S) -> Result<(), SignatureError>;
+
+    fn decoded_verify(&self, msg: &[u8], data: String) -> Result<(), SignatureError> {
+        let decoded_sig = self.decode(data)?;
+        return self.verify(msg, &decoded_sig);
+    }
+
+    fn decoded_relational_verify(
+        &self,
+        msg: &[u8],
+        data: String,
+        relation: VerificationRelation,
+    ) -> Result<(), SignatureError>;
+
+    fn relational_verify(
+        &self,
+        msg: &[u8],
+        signature: &S,
+        relation: VerificationRelation,
+    ) -> Result<(), SignatureError>;
+
+    fn decode(&self, encoded_sig: String) -> Result<S, SignatureError>;
+}
+
+pub struct Ed25519DidVerifier {
+    public_key: ed25519_zebra::VerificationKey,
+}
+
+impl From<&crate::signer::Ed25519DidSigner> for Ed25519DidVerifier {
+    fn from(signer: &crate::signer::Ed25519DidSigner) -> Self {
+        Self {
+            public_key: signer.public_key,
+        }
+    }
+}
+
+impl DIDVerifier<Ed25519Signature> for Ed25519DidVerifier {
+    fn verify(&self, msg: &[u8], sig: &Ed25519Signature) -> Result<(), SignatureError> {
+        let sig_bytes: [u8; 64] = sig.0.as_slice().try_into().unwrap();
```

Comment:
Careful of using .`unwrap()` here, do not want to core dump.

Reply:
Maybe worth adding a `thiserror` enum to `SignatureError` or keeping a higher-level `Error` type that encapsulates all error cases:

```rust
#[error(transparent)]
TryFromSlice(#[from] std::array::TryFromSliceError),
```

then you use the `?` operand, e.g. `let sig_bytes: [u8; 64] = sig.0.as_slice().try_into()?`

Reply:
@Ryanmtate  I don't think that a rust panic [core dumps by default](https://dev.to/jcarlosv/core-dump-for-rust-10nm). I may be wrong though.